components/core/src/clp_s/ReaderUtils.cpp

```diff
@@ -180,51 +180,84 @@ std::vector<int32_t> ReaderUtils::get_schemas(std::string const& archive_path) {
+            break;
+        case NodeType::DATESTRING:
+            column_reader = new DateStringColumnReader(key_name, column_id, timestamp_dict);
+            break;
+        case NodeType::FLOATDATESTRING:
+            column_reader = new FloatDateStringColumnReader(key_name, column_id);
+            break;
+        case NodeType::OBJECT:
+        case NodeType::NULLVALUE:
+            reader->append_column(column_id);
+            break;
+    }
+
+    if (column_reader) {
+        reader->append_column(column_reader);
+    }
+    return column_reader;
+}
+
 void ReaderUtils::append_reader_columns(
         SchemaReader* reader,
         std::set<int32_t>& columns,
         std::shared_ptr<SchemaTree> const& schema_tree,
         std::shared_ptr<VariableDictionaryReader> const& var_dict,
         std::shared_ptr<LogTypeDictionaryReader> const& log_dict,
         std::shared_ptr<LogTypeDictionaryReader> const& array_dict,
-        std::shared_ptr<TimestampDictionaryReader> const& timestamp_dict
+        std::shared_ptr<TimestampDictionaryReader> const& timestamp_dict,
+        bool extract_timestamp
 ) {
-    for (int32_t column : columns) {
-        auto node = schema_tree->get_node(column);
-        std::string key_name = node->get_key_name();
-        switch (node->get_type()) {
-            case NodeType::INTEGER:
-                reader->append_column(new Int64ColumnReader(key_name, column));
-                break;
-            case NodeType::FLOAT:
-                reader->append_column(new FloatColumnReader(key_name, column));
-                break;
-            case NodeType::CLPSTRING:
-                reader->append_column(
-                        new ClpStringColumnReader(key_name, column, var_dict, log_dict)
-                );
-                break;
-            case NodeType::VARSTRING:
-                reader->append_column(new VariableStringColumnReader(key_name, column, var_dict));
-                break;
-            case NodeType::BOOLEAN:
-                reader->append_column(new BooleanColumnReader(key_name, column));
-                break;
-            case NodeType::ARRAY:
-                reader->append_column(
-                        new ClpStringColumnReader(key_name, column, var_dict, array_dict, true)
-                );
-                break;
-            case NodeType::DATESTRING:
-                reader->append_column(new DateStringColumnReader(key_name, column, timestamp_dict));
-                break;
-            case NodeType::FLOATDATESTRING:
-                reader->append_column(new FloatDateStringColumnReader(key_name, column));
-                break;
-            case NodeType::OBJECT:
-            case NodeType::NULLVALUE:
-                reader->append_column(column);
-                break;
+    int32_t timestamp_column_id = -1;
+    if (timestamp_dict->has_timestamp()) {
+        timestamp_column_id = timestamp_dict->get_first_column_id();
```

Comment:
I can see what you're trying to do with the changes to the timestamp dictionary, but only allowing the first column Id in the timestamp dictionary to be the authoritative timestamp column is a bug.

The current design allows multiple mst nodes (with different types, but the same key name) to correspond to the timestamp-key passed at ingestion time.

Logs like the following ingested with --timestamp-key ts
```
{"ts": 1234}
{"ts": 1235.0}
{"ts": "2015-01-31T15:50:45.392"}
```
are perfectly valid, and require extracting the timestamp from a column with a different mst node Id in each schema.

To avoid resolving this per-schema during Schema Matching we should probably use a set of mst node Ids per timestamp key name and then just check for set membership in the code below.

Reply:
To be a bit more specific I'm suggesting we encode each timestamp entry as `<# mst node ids> <mst node id 0> ... <mst node id N-1> <strlen> <name>`

Reply:
Do you think we should pass both `std::string key` and `int32_t id` to `ingest_entry` and add the timestamp entry and id to two mappings `std::map<std::string, TimestampEntry>` and `std::map<std::string, std::set<int32_t>>`, or we only pass `int32_t id` to `ingest_entry` and merge ranges with the same key name but different ids when writing to disk?

Reply:
I suppose merging at write time is more efficient, so lets do that. Potentially more bug prone, but we only have to get it right once.